README.md
# Colonial Wars Main
This is ``colonialwars-main``, the component that makes Colonial Wars run.
This repository includes:
- A class (``ColonialwarsManager``) to manage the execution of Colonial Wars
App server and Colonial Wars Game server processes; and,
- Files as entrypoints for the application. These are found in the ``/bin`` directory.

## Running the full app
### Running from source
To run the full Colonial Wars application from source, you will need:
- ``colonialwars-main`` (this repository), to manage the execution of the other
components;
- ``colonialwars-appserver``, to run the Colonial Wars application server that
will serve non-game server routes; and,
- ``colonialwars-gameserver``, to run the Colonial Wars game server that will
run the actual game.

Those repositories need to be put in this directory format:
```none
[project-root]/
 |-- colonialwars-appserver/
 |-- colonialwars-gameserver/
 |-- colonialwars-main/
```
Then, you must ``cd`` into ``colonialwars-main``, and run ``npm start`` (but make sure
you have all the dependencies installed first).
### Running from binaries
Currently not available.

## Compatibility
Compatible with ``colonialwars-appserver`` v0.1.x.

## Running the tests
Download this repository somehow, ``cd`` into the project root, and run:
```sh
npm install
npm test
```
(just like any other Node.JS project.)


CHANGELOG.md
# Colonial Wars Main Changelog
Changelog for ``colonialwars-main``.

The format is based on [Keep a Changelog][1], and this project adheres to [Semantic Versioning][2].

## [v0.1.2] - 2020-12-17
### Added:
- Added ``winston`` and ``winston-syslog`` to do application logging.
### Changed:
- Changed expectations when testing classes–now we also check if a value is an *instance*
of a class, instead of just checking if the error was ``null``.
- Made some micro-optimizations... Meaning, I replaced a few loops here, tried not to access
the same property of an object repetitively in a loop, etc.
- Made class properties more obvious, by setting uninitialized class properties to ``null``,
instead of just letting people find where they're defined.
- Moved all graceful shutdown handlers to other files.
- Stopped passing ``DEBUG`` environment variable if ``IS_PROD`` is set to ``true``.

## [v0.1.1] - 2020-12-13
### Changed:
- Made ``ColonialwarsManager`` clear all objects that store information about was-running
subprocesses after calls to ``.stopAll()``.
- Made this package pass down the ``DEBUG`` environment variable to subprocesses.
- Used ``error.stack`` instead of the ``error`` object.
- Used the graceful shutdown handlers of the ``AppServer`` class (from ``colonialwars-appserver``)
in ``app-bin.js``.
### Fixed:
- Fixed broken CHANGELOG links.

## [v0.1.0] - 2020-12-12
- Initial (pre-)release

[1]: https://keepachangelog.com/
[2]: https://semver.org

[v0.1.0]: https://github.com/Take-Some-Bytes/colonialwars-main/tree/2ca231664a1ef41334d8bd500f0fddf08654310e
[v0.1.1]: https://github.com/Take-Some-Bytes/colonialwars-main/tree/025bd6b1accd68e752a5a5fbc74a475451c94f0d
[v0.1.2]: https://github.com/Take-Some-Bytes/colonialwars-main/tree/main


spec/utils_spec.js
/* eslint-env jasmine */
/**
 * @fileoverview Testing application utility methods.
 */

const utils = require('../lib/utils/utils')

describe('The Colonial Wars main utility object,', () => {
  it('should have four properties', () => {
    const utilsLength = Object.keys(utils).length

    expect(utilsLength).toBe(4)
  })
  it('should have two functions, one symbol, one number', () => {
    const utilVals = Object.values(utils)
    let numFuncs = 0
    let numSymbols = 0
    let numNums = 0
    utilVals.forEach(val => {
      switch (typeof val) {
        case 'function':
          numFuncs++
          break
        case 'symbol':
          numSymbols++
          break
        case 'number':
          numNums++
          break
      }
    })

    expect(numFuncs).toBe(2)
    expect(numSymbols).toBe(1)
    expect(numNums).toBe(1)
  })
  it('should be able to deep-freeze objects', () => {
    const obj = {
      foo: 'bar',
      baz: 'quaxx',
      cheese: 'I don\'t like it.',
      deep: {
        deepdeep: {
          depth: 'deeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeep'
        }
      }
    }
    const errors = []
    try {
      utils.deepFreeze(obj)
      errors.push(null)
    } catch (ex) {
      errors.push(ex)
    }
    Object.keys(obj).forEach(key => {
      try {
        obj[key] = null
      } catch (ex) {
        errors.push(ex)
      }
    })

    expect(errors.shift()).toBe(null)
    expect(errors.every(val => val instanceof TypeError)).toBe(true)
  })
  it('should be able to generate IDs', async () => {
    console.log()
    const idLength = 36
    const ids = await Promise.all(
      Array.apply(null, { length: 100 }).map(val => {
        return (val = utils.genID(idLength))
      })
    )
    const idSet = new Set(ids)
    // console.dir(ids)
    // console.dir(idSet)
    expect(ids.length).toBe(idSet.size)
  })
})


spec/config_spec.js
/* eslint-env jasmine */
/**
 * @fileoverview Test Configurations class.
 */

const path = require('path')
const Configurations = require('../lib/utils/configurations')

describe('The Configurations class, when used without a file input,', () => {
  let configurations = null
  let err = null

  it('should construct without error', () => {
    try {
      configurations = new Configurations({
        crud: [
          'cREATE', 'rEAD', 'uPDATE', 'dELETE'
        ],
        data: {
          trash: true,
          recyclable: false,
          message: "Don't recycle this trash."
        }
      })
    } catch (ex) {
      err = ex
    }

    expect(err).toBe(null)
    expect(configurations).toBeInstanceOf(Configurations)
  })

  it('should get existing properties and return them', () => {
    let val = null
    console.log()
    if (configurations instanceof Configurations) {
      val = configurations.get('crud')
      console.log(val)
    }
    expect(val).not.toBe(null)
  })

  it('should return null for non-existing properties', () => {
    let val = null
    console.log()
    if (configurations instanceof Configurations) {
      val = configurations.get('i', "don't", 'exist')
      console.log(val)
    }
    expect(val).toBe(null)
  })
})

describe('The Configurations class, when used with a file input,', () => {
  let configurations = null
  let err = null

  it('should create without error', async () => {
    try {
      configurations = await Configurations.readFrom(
        path.join(__dirname, 'mocks/external', 'mock-config.json')
      )
    } catch (er) {
      err = er
    }
    expect(err).toBe(null)
    expect(configurations).toBeInstanceOf(Configurations)
  })

  it('should get existing properties and return them', () => {
    let vals = null
    console.log()
    if (configurations instanceof Configurations) {
      vals = []
      vals.push(configurations.get('This'))
      console.dir(vals)
      vals.push(configurations.get('mock'))
      console.dir(vals)
    }
    expect(vals).not.toBe(null)
    expect(vals).toHaveSize(2)
  })

  it('should return null for non-existing properties', () => {
    let val = null
    console.log()
    if (configurations instanceof Configurations) {
      val = configurations.get('i', 'still', "don't", 'exist')
      console.log(val)
    }
    expect(val).toBe(null)
  })
})

describe('The Configurations class, when used with fallbacks,', () => {
  let configurations = null
  let err = null

  it('should create without error, even if the file path is not valid', async () => {
    try {
      configurations = await Configurations.readFrom(
        '',
        { fellback: true, That: 'Not this, THAT' }
      )
    } catch (er) {
      err = er
    }
    expect(err).toBe(null)
    expect(configurations).toBeInstanceOf(Configurations)
  })

  it('should get existing properties and return them', () => {
    let vals = null
    console.log()
    if (configurations instanceof Configurations) {
      vals = []
      vals.push(configurations.get('That'))
      console.dir(vals)
      vals.push(configurations.get('fellback'))
      console.dir(vals)
    }
    expect(vals).not.toBe(null)
    expect(vals).toHaveSize(2)
  })

  it('should return null for non-existing properties', () => {
    let val = null
    console.log()
    if (configurations instanceof Configurations) {
      val = configurations.get('i', 'am', 'a', 'teapot')
      console.log(val)
    }
    expect(val).toBe(null)
  })
})


spec/logger_spec.js
/* eslint-env jasmine */
/**
 * @fileoverview Tests for Logger class.
 */

const winston = require('winston')
const Loggers = require('../lib/logging/loggers')
const MockSyslogServer = require('./mocks/external/mock-syslog-server')

/**
 * Promise delay.
 * @param {number} time The time to delay for.
 */
const delay = (time) => new Promise(resolve => setTimeout(resolve, time))

describe('The Loggers class, when used without a Syslog server,', () => {
  const oldProcessTitle = process.title
  let loggers = null
  let err = null

  beforeAll(() => {
    // Make sure process.title is `testing_logger`.
    process.title = 'testing_logger'
  })
  afterAll(() => {
    // Reset process title.
    process.title = oldProcessTitle
  })

  it('should construct without error', () => {
    try {
      loggers = new Loggers({
        isProd: false,
        debug: console.log,
        loggerInfos: [
          {
            id: 'Process-logger',
            label: 'Process_log'
          }
        ],
        // Use syslog config.
        levels: winston.config.syslog.levels,
        colors: winston.config.syslog.colors
      })
    } catch (ex) {
      err = ex
    }

    expect(err).toBe(null)
    expect(loggers).toBeInstanceOf(Loggers)
  })

  it('should have one logger', () => {
    let numLoggers = 0
    if (loggers instanceof Loggers) {
      numLoggers = loggers.allLoggers().length
    }
    expect(numLoggers).toBe(1)
  })

  it('should be able to log (with colour)', () => {
    let logged = 0
    if (loggers instanceof Loggers) {
      loggers.allLoggers().forEach(logger => {
        console.log()
        logger.debug('DEBUG level')
        logger.info('INFO level')
        logger.notice('NOTICE level')
        logger.warning('WARNING level')
        logger.error('ERROR level')
        logger.crit('CRITICAL level')
        logger.alert('ALERT level')
        logger.emerg('EMERGENCY level')
        logged++
      })
    }
    expect(logged).toBeGreaterThanOrEqual(1)
  })
})

describe('The Loggers class, when used with a mock Syslog server,', () => {
  const syslogServer = new MockSyslogServer()
  let loggers = null
  let err = null

  // First, start the mock Syslog server.
  beforeAll(async () => {
    await syslogServer.start()
  })
  // Clean up the server once done.
  afterAll(async () => {
    await syslogServer.stop()
  })

  it('should construct without error', () => {
    console.log()
    try {
      loggers = new Loggers({
        isProd: true,
        debug: console.log,
        loggerInfos: [
          {
            id: 'Process-logger',
            label: 'Process_log'
          }
        ],
        // Use syslog config
        levels: winston.config.syslog.levels,
        colors: winston.config.syslog.colors,
        syslogOpts: {
          port: 5514,
          host: 'localhost',
          protocol: 'tcp4',
          type: 'rfc5424',
          eol: '\r\n'
        }
      })
    } catch (ex) {
      err = ex
    }
    expect(err).toBe(null)
  })

  it('should be able to send logs to syslog server', async () => {
    let dataReceived = false
    console.log()
    if (loggers instanceof Loggers) {
      loggers.allLoggers().forEach(logger => {
        logger.alert('If you don\'t log this you don\'t work.')
      })
    }
    await delay(1000)
    dataReceived = syslogServer.dataReceived
    expect(dataReceived).toBe(true)
  })
})


lib/utils/utils.js
/**
 * @fileoverview Utility functions. There's no reason to wrap this
 * in a class... Right? I believe so.
 */

const NO_FREEZE = Symbol('NO_FREEZE')
const ID_MAX_LEN = 100
const crypto = require('crypto')

/**
 * Deep-freezes an object.
 * @param {Object} object The object to deep freeze.
 * @returns {Object}
 */
function deepFreeze (object) {
  // Retrieve the property names defined on object.
  const propNames = Object.getOwnPropertyNames(object)
  // Check for a special "NO_FREEZE" symbol.
  if (object[NO_FREEZE] === true) {
    return
  }
  // Freeze properties before freezing self.
  for (const name of propNames) {
    const value = object[name]
    if (value && typeof value === 'object') {
      // Check for a special "NO_FREEZE" symbol.
      if (value[NO_FREEZE] === true) {
        return
      }
      deepFreeze(value)
    }
  }
  return Object.freeze(object)
}
/**
 * Binds a number to the given minimum and maximum, inclusive of both
 * binds. This function will still work if min and max are switched.
 * @param {number} val The value to check.
 * @param {number} min The minimum number to bound to.
 * @param {number} max The maximum number to bound to.
 * @returns {number}
 */
function bind (val, min, max) {
  if (min > max) { return Math.min(Math.max(val, max), min) }
  return Math.min(Math.max(val, min), max)
}
/**
 * Mixes up a string real good.
 * @param {string} string The string to mix up.
 * @returns {string}
 */
function mixUp (string) {
  const splitString = [...string]
  const strLen = string.length
  let result = ''

  for (let i = 0; i < strLen; i++) {
    const indexToGet = bind(
      Math.floor(Math.random() * splitString.length),
      0, splitString.length
    )
    result += splitString.splice(indexToGet, 1)[0]
    i++
  }

  return result
}
/**
 * Generates an ID that is quite unique.
 * @param {number} length The length of the ID.
 * @returns {Promise<string>}
 */
function genID (length) {
  return new Promise((resolve, reject) => {
    if (length > ID_MAX_LEN) {
      reject(new RangeError('Requested ID is too large!'))
      return
    }

    const cryptoBytes = Math.round(
      length - (length / 4)
    )
    const randChars = length / 4
    const chars =
      'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_'
    const charsLen = chars.length
    let id = ''

    crypto.randomBytes(cryptoBytes, (err, buf) => {
      if (err) {
        reject(err)
        return
      }
      try {
        id = buf.toString('hex')
        for (let i = 0; i < randChars; i++) {
          id += chars.charAt(Math.floor(Math.random() * charsLen))
        }
        id = mixUp(id)
        resolve(id)
      } catch (err) {
        reject(err)
      }
    })
  })
}

module.exports = exports = {
  // Symbols:
  NO_FREEZE,
  // Constants
  ID_MAX_LEN,
  // Functions:
  deepFreeze,
  genID
}


lib/logging/loggers.js
/**
 * @fileoverview Loggers class.
 */

const winston = require('winston')
const winstonSyslog = require('winston-syslog')
const Syslog = winstonSyslog.Syslog

/**
 * @typedef {Object} WinstonTransports
 * @prop {winston.transports.ConsoleTransportInstance|null} consoleTransport
 * @prop {Syslog|null} syslogTransport
 *
 * @typedef {Object} LoggerInfo
 * @prop {string} id
 * @prop {string} label
 *
 * @typedef {Object} LoggingConfig
 * @prop {boolean} isProd
 * @prop {import('debug').Debugger} debug
 * @prop {Array<LoggerInfo>} loggerInfos
 * @prop {winstonSyslog.SyslogTransportOptions} syslogOpts
 * @prop {winston.config.AbstractConfigSetLevels} levels
 * @prop {Object<string, string>} colors
 *
 * @typedef {Object} LoggersInterface
 * @prop {import('debug').Debugger} debug
 * @prop {(id: string) => winston.Logger} get
 */

/**
 * Loggers class.
 */
class Loggers {
  /**
   * Constructor for a Loggers class.
   * @param {LoggingConfig} config Configurations.
   */
  constructor (config) {
    const {
      isProd, loggerInfos,
      debug, syslogOpts,
      levels, colors
    } = config

    // Keep an array of logger IDs so that we could get
    // all the loggers stored here easily.
    this.loggerIDs = []

    this.debug = debug
    this.loggers = new winston.Container()
    this.logLevels = levels
    this.logColors = colors
    this.logFormat = winston.format.printf(info => {
      // Use process.title just to be much clearer what process
      // these logs are originating from.
      return (
        `${process.title}: ${info.timestamp}` +
        ` [${info.label}] ${info.level}: ${info.message}`
      )
    })
    this.init(isProd, loggerInfos, syslogOpts)
  }

  /**
   * Creates the winston transports.
   * @param {boolean} isProd Whether the app is in production or not.
   * @param {winstonSyslog.SyslogTransportOptions} [syslogOpts] Syslog options,
   * if in production.
   * @returns {WinstonTransports}
   * @private
   */
  _createTransports (isProd, syslogOpts) {
    const consoleTransport =
      new winston.transports.Console({
        // I want to use CRLF.
        eol: '\r\n',
        level: 'debug'
      })
    if (isProd) {
      if (
        !syslogOpts ||
        typeof syslogOpts !== 'object' ||
        typeof syslogOpts.constructor !== 'function' ||
        syslogOpts.constructor !== Object
      ) {
        throw new TypeError(
          'If production mode is enabled, must pass syslog options!'
        )
      }
      this.debug(
        `Created syslog transport with options ${JSON.stringify(syslogOpts, null, 2)}`
      )
      return {
        consoleTransport: null,
        syslogTransport: new Syslog(syslogOpts)
      }
    }
    this.debug('Returning console only transports.')
    return {
      consoleTransport: consoleTransport,
      syslogTransport: null
    }
  }

  /**
   * Initializes this Loggers class.
   * @param {boolean} isProd Whether the app is in production or not.
   * @param {Array<LoggerInfo>} loggerInfos Logger info.
   * @param {winstonSyslog.SyslogTransportOptions} syslogOpts Syslog options,
   * if in production.
   */
  init (isProd, loggerInfos, syslogOpts) {
    const transports = this._createTransports(isProd, syslogOpts)
    winston.addColors(this.logColors)
    loggerInfos.forEach(loggerInfo => {
      // Use console transports only if not in production.
      const currentTransports = [transports.consoleTransport]
      if (isProd) {
        currentTransports.shift()
        currentTransports.push(transports.syslogTransport)
      }
      this.loggerIDs.push(loggerInfo.id)
      this.loggers.add(loggerInfo.id, {
        levels: this.logLevels,
        transports: currentTransports,
        format: winston.format.combine(
          winston.format.colorize({
            colors: this.logColors
          }),
          winston.format.label({ label: loggerInfo.label }),
          winston.format.timestamp(),
          this.logFormat
        )
      })
    })
    this.debug('All loggers created!')
  }

  /**
   * Gets the specified logger.
   * @param {string} loggerId The logger ID.
   * @returns {winston.Logger}
   */
  get (loggerId) {
    // Literately just a proxy method.
    return this.loggers.get(loggerId)
  }

  /**
   * Returns all the loggers that this class has.
   * @returns {Array<winston.Logger>}
   */
  allLoggers () {
    const loggers = []
    this.loggerIDs.forEach(id => {
      loggers.push(this.get(id))
    })
    return loggers.filter(logger => typeof logger !== 'undefined' && logger !== null)
  }
}

module.exports = exports = Loggers


lib/utils/configurations.js
/**
 * @fileoverview Configurations class to read application configurations.
 */

const fs = require('fs').promises
const utils = require('./utils')

/**
 * Configurations class for safe accessing of configurations.
 */
class Configurations {
  /**
   * Constructor for a Configurations class. You may only set configurations
   * one time. After the first time, it is read-only.
   * @param {Object<string, any>} config The configurations that was received
   * from user input.
   */
  constructor (config) {
    this.config = config
    this._deepFreeze()
  }

  /**
   * Gets the property specified. Returns `null` if
   * property does not exist.
   * @param  {...string} propNames Property names.
   * @returns {any}
   */
  get (...propNames) {
    let val = null
    for (const prop of propNames) {
      // Get the next value.
      val = typeof val === 'object' && val !== null
        ? val[prop]
        : this.config[prop]

      // If val is undefined, return null. Why null?
      // So that we know "nothing" has been returned intentionally.
      if (val === undefined) {
        return (val = null)
      }
    }
    return val
  }

  /**
   * Deep freezes configurations object.
   * @private
   */
  _deepFreeze () {
    if (Object.isFrozen(this.config)) { return }
    utils.deepFreeze(this.config)
  }

  /**
   * Reads a JSON configuration from the specified path.
   * @param {string} path The path of the configurations.
   * @param {Object} [fallbacks] Fallbacks if the config file could not be read.
   * @returns {Promise<Configurations>}
   */
  static async readFrom (path, fallbacks) {
    let contents = null
    try {
      contents = await fs.readFile(path)
    } catch (err) {
      // Only throw the error if no fallbacks were specified.
      if (fallbacks) {
        contents = JSON.stringify(fallbacks)
      } else {
        throw new Error(
          'Error occured while reading configurations!' +
          `Error: ${err.stack}`
        )
      }
    }
    const parsedContents = JSON.parse(contents.toString('utf-8'))
    return new Configurations(parsedContents)
  }
}

module.exports = exports = Configurations


lib/logging/loggers-config.js
/**
 * @fileoverview LoggersConfig class to manage configurations for the Loggers class.
 */

/**
 * @typedef {Object<string, any>} Config
 */

/**
 * LoggersConfig class.
 */
class LoggersConfig {
  constructor () {
    this.prodConfig = {}
    this.devConfig = {}
  }

  /**
   * Sets either the ``'prod'`` configurations, or the ``'dev'`` configurations.
   * @param {'prod'|'dev'} which Which config to set.
   * @param {Config} config The contents of the config.
   */
  setConfig (which, config) {
    // First, some type checking.
    if (which !== 'dev' && which !== 'prod') {
      throw new TypeError(
        'The `which` parameter must either `prod` or `dev`!'
      )
    } else if (!config || typeof config !== 'object') {
      throw new TypeError(
        'The `config` parameter must be a plain object!'
      )
    }

    // Now that type checks have passed, we could proceed
    // safely (I hope).
    this[`${which}Config`] = config
  }

  /**
   * Gets either the ``'prod'`` or the ``'dev'`` configurations.
   * @param {'prod'|'dev'} which Which config to get.
   * @returns {Config}
   */
  getConfig (which) {
    // First, some type checking.
    if (which !== 'dev' && which !== 'prod') {
      throw new TypeError(
        'The `which` parameter must either `prod` or `dev`!'
      )
    }
    // Now, PROCEED.
    return this[`${which}Config`]
  }
}

module.exports = exports = LoggersConfig


lib/colonialwars-manager.js
/**
 * @fileoverview ColonialwarsManager class to manage
 * the Colonial Wars game.
 */

const childProcess = require('child_process')
const winstonConfig = require('winston/lib/winston/config')
const path = require('path')
const debug = require('debug')

const utils = require('./utils/utils')
const constants = require('./constants')
const Loggers = require('./logging/loggers')
const Configurations = require('./utils/configurations')
const LoggersConfig = require('./logging/loggers-config')

/**
 * @typedef {Object} CWProcess
 * @prop {childProcess.ChildProcess} proc
 * @prop {string} id
 */

/**
 * ColonialwarsManager class.
 */
class ColonialwarsManager {
  /**
   * Constructor for a ColonialwarsManager.
   * @param {string} confFile Path to the configuration file.
   */
  constructor (confFile) {
    this.confFile = confFile || ''
    this.debug = debug('colonialwars-main')

    /**
     * @type {Map<string, childProcess.ChildProcess>}
     */
    this.appServers = new Map()
    /**
     * @type {Array<CWProcess>}
     */
    this.processes = []

    this.config = null
    this.appBin = null
    this.maxApps = null
    this.loggers = null
    this.loggersConf = null
  }

  /**
   * Gets self configurations.
   * @private
   */
  async _getConfig () {
    this.config = await Configurations.readFrom(
      this.confFile, constants.FALLBACKS
    )
    this.debug('Got configuration file.')

    // Set the NODE_ENV.
    process.env.NODE_ENV = this.config.get('IS_PROD')
      ? 'production'
      : 'development'
    this.appBin =
      this.config.get('APP_BIN_LOCATION') ||
      path.join(__dirname, '../bin', 'app-bin/app-bin.js')
    this.maxApps =
      typeof this.config.get('MAX_APPS') === 'number'
        ? this.config.get('MAX_APPS')
        : 1

    this.debug(
      'All other configurations determined'
    )
  }

  /**
   * Gets logging configurations.
   * @private
   */
  _getLoggersConfig () {
    const conf = {
      isProd: false,
      debug: this.debug,
      loggerInfos: [
        {
          id: 'Process-logger',
          label: 'Process_log'
        },
        {
          id: 'Manager-logger',
          label: 'Manager_log'
        }
      ],
      levels: winstonConfig.syslog.levels,
      colors: winstonConfig.syslog.colors,
      syslogOpts: null
    }
    this.loggersConf = new LoggersConfig()
    this.loggersConf.setConfig('dev', conf)
    this.loggersConf.setConfig('prod', Object.assign(
      Object.assign({}, conf), {
        isProd: true,
        syslogOpts: this.config.get('SYSLOG_OPTS')
      }
    ))
  }

  /**
   * Initialize helper classes.
   * @private
   */
  _initHelpers () {
    this.loggers = new Loggers(
      this.config.get('IS_PROD')
        ? this.loggersConf.getConfig('prod')
        : this.loggersConf.getConfig('dev')
    )
    this.debug('Initialized manager loggers.')
  }

  /**
   * Initializes this ColonialwarsManager.
   */
  async init () {
    await this._getConfig()
    this._getLoggersConfig()
    this._initHelpers()

    this.loggers.get('Manager-logger').info(
      'Initialization complete.'
    )
  }

  // Managing functions.
  /**
   * Adds a new colonialwars-appserver process. Returns the created
   * app server process.
   * @param {string} [confFile] Path of the configuration file.
   * @returns {Promise<CWProcess>}
   */
  async addAppProcess (confFile) {
    if (this.appServers.size >= this.maxApps) {
      throw new Error(
        'Maximum app server amount reached!'
      )
    }
    const appProc = childProcess.fork(
      this.appBin, [
        '--confFile', this.config.get('APP_CONF_FILE') || confFile || ''
      ],
      { env: { DEBUG: process.env.DEBUG } }
    )
    const appID = await utils.genID(36)
    this.appServers.set(appID, appProc)
    this.processes.push({
      proc: appProc,
      id: appID
    })

    this.loggers.get('Process-logger').info(
      `App ID ${appID} started with PID ${appProc.pid}.`
    )

    return {
      proc: appProc,
      id: appID
    }
  }

  /**
   * Stops all processes this manager is managing with the specified
   * signal. Default signal is SIGINT.
   * @param {NodeJS.Signals} [signal]
   */
  stopAll (signal) {
    if (!signal || typeof signal !== 'string') {
      signal = 'SIGINT'
    }

    this.processes.forEach(proc => {
      /**
       * TODO: Find out how to know if a subprocess has been terminated.
       * If apps are stuck in a infinite loop and have no time to pay attention
       * to handle-able signals, we might have to kill the process with SIGKILL.
       * But, there is no way to check if the process has terminated or not.
       * (2020/12/17) Take-Some-Bytes */
      proc.proc.kill(signal)

      this.loggers.get('Process-logger').info(
        `Stopped process ${proc.id} with signal ${signal}.`
      )
    })
    // Clear all the objects that store information about
    // the was-running processes.
    this.processes.length = 0
    this.appServers.clear()

    this.loggers.get('Process-logger').info(
      `All processes stopped with signal ${signal}.`
    )
  }

  // Static functions.
  /**
   * Creates a ColonialwarsManager. This is mainly here so that this class
   * could be initialized on creation.
   * @param {string} [confFile] The configuration file path.
   * @returns {Promise<ColonialwarsManager>}
   */
  static async create (confFile) {
    const cwManager = new ColonialwarsManager(confFile)
    await cwManager.init()
    return cwManager
  }
}

module.exports = exports = ColonialwarsManager
